[DX-1015,DX-1036] Create Tyk Charts Product Stack Section (Creates new helm chart structure in Product Stack section) (#4076)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-charts/tyk-control-plane-chart.md

---

> title: Tyk Control Plane Chart

Reviewer: Updated and ready for review

    # The license key needed by Tyk Dashboard to work.
    #
    # NOTE: If you do not want to store license as a plain text in the file, you can use a Kubernetes secret
    # that stores the dashboard license. Please see `.global.secrets.useSecretName`.
    dashboard: ""
```

#### Enabling Dashboard TLS

Assuming that TLS certificates for the Tyk Dashboard are available in the Kubernetes Secret `tyk-dashboard-tls`, follow these steps to enable TLS:

1. Set `global.tls.dashboard` to `true`.
2. Set `tyk-dashboard.dashboard.tls.secretName` to the name of the Kubernetes secret containing TLS certificates for the Tyk Dashboard, in this case, `tyk-dashboard-tls`.
3. Define certificate configurations in `tyk-dashboard.dashboard.tls.certificates`, which generates `TYK_DB_HTTPSERVEROPTIONS_CERTIFICATES` for the Tyk Dashboard.

Optional Steps, if needed:
- Modify the secret mount path on the Tyk Dashboard Pod via `tyk-dashboard.dashboard.tls.certificatesMountPath`.
- If necessary, either enable `insecureSkipVerify` via `tyk-dashboard.dashboard.tls.certificates`, or mount CA information through `tyk-dashboard.dashboard.extraVolumes` and `tyk-dashboard.dashboard.extraVolumeMounts`.

> - If the `tyk-bootstrap` chart is used to bootstrap the Tyk Dashboard, ensure that it has certificates to send requests to the Tyk Dashboard or enable `insecureSkipVerify` in the `tyk-bootstrap` chart.

Reviewer 2: ```suggestion
- If the `tyk-bootstrap` chart is used to bootstrap the Tyk Dashboard, ensure that bootstrap app can validate certificate of Tyk Dashboard or enable `insecureSkipVerify` in the `tyk-bootstrap` chart.
```

3. [Tyk Dashboard License](#tyk-dashboard-license-required)
4. [Tyk MDCB License](#tyk-mdcb-license-required)

If you would like to use Developer Portal, additional license is required:

5. [Tyk Developer Portal License](#tyk-developer-portal-license-required)

Then just run:
```bash
helm install tyk-control-plane tyk-helm/tyk-control-plane -n tyk --create-namespace -f values.yaml --devel
```

### Uninstalling The Chart

```bash
helm uninstall tyk-control-plane -n tyk
```
This removes all the Kubernetes components associated with the chart and deletes the release.

### Upgrading Chart

```bash
helm upgrade tyk-control-plane tyk-helm/tyk-control-plane -n tyk -f values.yaml --devel
```

## Configuration

To get all configurable options with detailed comments:

```bash
helm show values tyk-helm/tyk-control-plane --devel > values.yaml
```

You can update any value in your local `values.yaml` file and use `-f [filename]` flag to override default values during installation. 
Alternatively, you can use `--set` flag to set it in Tyk installation. See [Using Helm](https://helm.sh/docs/intro/using_helm/) for examples.

### Set Redis Connection Details (Required)

> Tyk uses Redis for distributed rate-limiting and token storage. You may use the Bitnami chart or Tyk's `simple-redis` to install chart for POC purpose.

Reviewer: updated

{{< /note >}}

Follow the notes from the installation output to get connection details and password.

```
  Redis(TM) can be accessed on the following DNS names from within your cluster:

    tyk-redis-master.tyk.svc.cluster.local for read/write operations (port 6379)
    tyk-redis-replicas.tyk.svc.cluster.local for read-only operations (port 6379)

  export REDIS_PASSWORD=$(kubectl get secret --namespace tyk tyk-redis -o jsonpath="{.data.redis-password}" | base64 --decode)
```

The Redis address as set by Bitnami is `tyk-redis-master.tyk.svc.cluster.local:6379`

You can reference the password secret generated by Bitnami chart by  `--set global.redis.passSecret.name=tyk-redis` and `--set global.redis.passSecret.keyName=redis-password`, or just set `global.redis.pass=$REDIS_PASSWORD`

***Evaluation only: via *simple-redis* chart***

Another option for Redis, to get started quickly, is to use our [simple-redis](https://artifacthub.io/packages/helm/tyk-helm/simple-redis) chart.

{{< warning  success >}}
**Warning**

Please note that these provided charts must never be used in production or for anything

> but a quick start evaluation only. Use Bitnami Redis or Official Redis Helm chart in any other case.

Reviewer: updated

   # Default value: false
   # useSSL: false
```

**PostgresSQL Installation**
```bash
helm install tyk-postgres bitnami/postgresql --set "auth.database=tyk_analytics" -n tyk
```

Follow the notes from the installation output to get connection details.

{{< note success >}}
**Note**

Please make sure you are installing PostgreSQL versions that are supported by Tyk. Please refer to Tyk docs to get list of [supported versions]({{< ref "tyk-dashboard/database-options" >}}).
{{< /note >}}

```yaml
global:
  # Postgres connection string parameters.
  postgres:
    # host corresponds to the host name of postgres
    host: tyk-postgres-postgresql.tyk.svc
    # port corresponds to the port of postgres
    port: 5432
    # user corresponds to the user of postgres
    user: postgres
    # password corresponds to the password of the given postgres user in selected database
    password:
    # database corresponds to the database to be used in postgres
    database: tyk_analytics
    # sslmode corresponds to if postgres runs in sslmode (https)
    sslmode: disable
    # Connection string can also be set using a secret. Provide the name of the secret and key below.

>     # connectionStringSecret:

Reviewer: @buraksekili can you make the necessary changes in the values.yaml file? Suggest to add instruction to uncomment the line below. Would this take precedence over the above values when uncommented?

{{< /note >}}

***Tyk Developer Portal AWS S3 Access Key ID***

In order to refer Tyk Developer Portal AWS S3 Access Key ID through Kubernetes secret,
please use `tyk-dev-portal.useSecretName`, where the secret should contain a key called 
`DevPortalAwsAccessKeyId`.

{{< note success >}}
**Note**

If `tyk-dev-portal.useSecretName` is in use, please add all keys mentioned to the secret.
{{< /note >}}

***Tyk Developer Portal AWS S3 Secret Access Key***

In order to refer Tyk Developer Portal connection string to the selected database through Kubernetes secret,
please use `tyk-dev-portal.useSecretName`, where the secret should contain a key called 
`DevPortalAwsSecretAccessKey`.

{{< note success >}}
**Note**

If `tyk-dev-portal.useSecretName` is in use, please add all keys mentioned to the secret.
{{< /note >}}

***Redis Password***

Redis password can also be provided via a secret. Store Redis password in Kubernetes secret and refer to this secret
via `global.redis.passSecret.name` and `global.redis.passSecret.keyName` field, as follows:

```yaml
global:  
  redis:
     passSecret:
       name: "yourSecret"
       keyName: "redisPassKey"

> ```

Reviewer: updated

***MongoDB or Postgres connection strings***

Storage connection strings can also be provided via a secret. Store the connection string in Kubernetes secret and 
refer to this secret via `global.{mongo,postgres}.connectionURLSecret.name` and `global.{mongo,postgres}.connectionURLSecret.keyName` field, 
as follows:

- MongoDB:
```yaml
global:  
  mongo:
    connectionURLSecret:
       name: "yourSecret"

>        keyName: "redisPassKey"

Reviewer: good spot. corrected.

```

- Postgres:
```yaml
global:
  postgres:
    connectionURLSecret:
       name: "yourSecret"

>        keyName: "redisPassKey"

Reviewer: good spot. corrected.

```

### Gateway Configurations

{{< note success >}}
**Note**

In Tyk control plane, Tyk Gateway acts as a management gateway that is used for creation of keys and certificates. It does not service API requests. It is important to ensure there is no public access to it and it must not be sharded (tagged) as it “belongs” to the whole Tyk installation.
{{< /note >}}

Configure below inside `tyk-gateway` section.

#### Update Tyk Gateway Version

> Set version of gateway at `tyk-gateway.gateway.image.tag`. You can find the list of version tags available from [Docker hub](https://hub.docker.com/u/tykio). Please check [Tyk Release notes]({{<ref "/release-notes">}}) carefully while upgrading or downgrading.

Reviewer: updated

### Tyk MDCB Configurations

#### Tyk MDCB License (Required)

Tyk MDCB requires a license to be set at `tyk-mdcb.mdcb.license`. This field is mandatory and must be configured.

To enhance security and avoid storing plaintext values for the MDCB license directly in the Helm value file,
an alternative approach is available. You can store the license in a Kubernetes Secret and reference it externally.
Set the license in the Kubernetes Secret and provide the secret's name through `tyk-mdcb.mdcb.useSecretName`.
The Secret must contain a key named `MDCBLicense`.

#### Tyk MDCB Listen Port

The `tyk-mdcb.mdcb.listenPort` field represents a RPC port which worker Tyk Gateways will connect to.
Setting `tyk-mdcb.mdcb.listenPort` field opens a port on MDCB container and MDCB service targets this port.
It is used to set `TYK_MDCB_LISTENPORT`

#### Tyk MDCB Health Check Port
The health check port for Tyk MDCB can be configurable via `tyk-mdcb.mdcb.probes.healthCheckPort` field.
This port lets MDCB allow standard health checks.

It also defines the path for liveness and readiness probes.
It is used to set TYK_MDCB_HEALTHCHECKPORT

### Tyk Bootstrap Configurations

> To enable bootstrapping, set `global.components.bootstrap` to `true`. It would run [tyk-k8s-bootstrap](https://github.com/TykTechnologies/tyk-k8s-bootstrap) to bootstrap `tyk-stack` and to create Kubernetes secrets that can be utilized in Tyk Operator and Tyk Developer Portal.

Reviewer 2: ```suggestion
To enable bootstrapping, set `global.components.bootstrap` to `true`. It would run [tyk-k8s-bootstrap](https://github.com/TykTechnologies/tyk-k8s-bootstrap) to bootstrap `tyk-control-plane` and to create Kubernetes secrets that can be utilized in Tyk Operator and Tyk Developer Portal.
```
